Simple Path Selection Policies

koflin opened the pull request.

koflin commented:
@marcfrei @tzaeschke I've implemented a basic version of policies comparable to PAN and JPAN. I've created a `scion_policy` struct that represents a policy. The policy struct must implement a `void (*filter)(struct scion_path_collection *paths)` that is used to sort and filter the paths fetched from the control server. Since, we do not implement selectors yet, the socket will always choose the first element of the filtered and sorted path collection. In contrast to the PAN and JPAN implementation, this filter function does not return a collection of paths, because it is supposed to sort and filter the path list in-place. There are mainly two reasons for this design choice:

1. This is more memory efficient, since the collection does not have to be duplicated.
2. It makes the memory management easier for the socket. When filtering the collection in-place, paths that are filtered out are automatically deallocated. Otherwise the socket would need to keep two lists: a) the original list it received when fetching the paths and b) the filtered list. For deallocation we would then need two different methods for freeing a path collection: one that frees the list and their values (paths themselves) and one that only frees the list itself.

I've also implemented two standard policies `scion_policy_highest_mtu` and `scion_policy_least_hops`, where the later is used as the default policy for new sockets. The socket policy can be changed with `scion_setsockpolicy()`. An example which shows how to implement a custom policy is in `examples/policy.c`.

I've also had to extend the `path_collection` and `linked_list` to support a sorting and filter function, which can be used to implement the policies. Currently, we use an in-place merge sort to implement the sort function. 

What do you think about the interface?

marcfrei commented:
In my opinion, `examples/policy.c` looks great and is very intuitive to understand.

koflin commented:
> In my opinion, `examples/policy.c` looks great and is very intuitive to understand.

Cool! If @tzaeschke also likes the interface, I could continue with completing the public docs for this and implement the Path Metadata extensions (latency, bandwidth, geo, link type, etc.). For this we would need to do the following:

- [ ] add [path metadata config files](https://docs.scion.org/en/latest/manuals/control.html#path-metadata) to the default topology in [scion-testnet](https://github.com/marcfrei/scion-testnet). Maybe someone already did this for the default topology, then we could just copy over the config files. Do you know this @marcfrei ?
- [ ] extend `scion_path_metadata` with the additional metadata fields
- [ ] fill in the additional metadata fields when constructing the path
- [ ] add some getters for the path to get this information
- [ ] implement some more policies like `min_latency` and `max_bandwidth`.

tzaeschke commented:
I think the in-place sorting is a good idea!
I think it solves the problem that policies may be used concurrently and that, at the same time, sockets must be able to choose a path other than the first one (e.g. if the first path causes an error, or if we want to use multipath).

koflin commented:
@marcfrei @tzaeschke  I've implemented the collection of the extended path metadata (latency, bandwidth, etc.) following the [implementation in Go](https://github.com/scionproto/scion/blob/master/private/path/combinator/staticinfo_accumulator.go) with some modifications. I've also implemented the two built in path selection policies `scion_policy_highest_bandwidth` and `scion_policy_lowest_latency` that PAN and JPAN implement. Important modifications include:

1. Implementation of a map data structure in `util/map.c` that is used in the metadata collection algorithm. For the sake of simplicity, the implementation uses a linked list to store the key-value pairs at the moment. In the future, we should use a hashmap implementation to be more efficient.
2. Extending and filling the `scion_as_entry` with extended metadata information returned from the control service in `common/as_entry.h` and `control_plane/segment.c` respectively.
3. Creation of a new file `control_plane/path_metadata` that contains the logic for gathering the extended metadata of a path.
4. I've created and exposed the metadata structs and types in `scion.h` and added a getter `scion_path_get_metadata()` that returns a pointer to the metadata of a path. This can then be used to implement custom path policies that use this metadata.
5. I've implemented and exposed a function `scion_path_collection_as_array()` that is used to get all the paths from a path collection in form of an array. This can then be used to iterate over the paths, which previously wasn't possible (in the public API).
6. Implementation of `scion_policy_highest_bandwidth` and `scion_policy_lowest_latency` in `control_plane/policy.c`.

With these modifications users of the library can now implement path selection policies that make use of the path metadata information such as latency, bandwidth, geo data, etc. What do you think?

tzaeschke commented on lib/include/scion/scion.h:
```diff
@@ -320,31 +320,151 @@ struct scion_path;
 
 /**
- * Gets the weight of a path.
+ * Gets the hops of a path.
```
Strictly speaking, it returns the hopcount, not the hops.

tzaeschke commented on lib/common/path_collection.c:
```diff
@@ -20,20 +20,30 @@
-	struct scion_path_collection *new_paths = malloc(sizeof(*paths));
-	if (new_paths == NULL) {
-		return SCION_MEM_ALLOC_FAIL;
+	struct scion_path_collection *paths = malloc(sizeof(*paths));
```
I'm a bit unsure about this.
According to this: https://stackoverflow.com/questions/2022335/whats-the-point-of-malloc0
`malloc(0)` may return a non-NULL address.

koflin replied:
`sizeof` returns the size of the `struct scion_path_collection` here, not the amount of elements in the the path collection

tzaeschke replied:
Right, ok, so what was wrong with returning `SCION_MEM_ALLOC_FAIL`?

koflin replied:
Actually, I don't even need this helper function anymore. I can move the code directly into the `scion_path_collection_init`.

tzaeschke commented on lib/data_plane/path.c:
```diff
@@ -238,85 +221,27 @@ void scion_path_print(const struct scion_path *path)
 }
 
-uint32_t scion_path_get_mtu(const struct scion_path *path)
+struct scion_path_metadata *scion_path_get_metadata(struct scion_path *path)
```
Should that maybe return a `const`? And the argument could be `const`?

koflin replied:
I agree that it would be very nice to make the return type of this function `const`. This would ensure that a user of the library only reads the metadata and does not accidentally modify it.

The argument can also be made `const` but is less usefull.

tzaeschke replied:
I forgot what the plan is for sharing paths between sockets. Do we copy paths or do we make them immutable? If it is the latter then making the argument const would be useful?
I guess the question is more whether we do it now or later?

tzaeschke replied:
The idea would be to prevent accidental modification, which would be harmful with concurrent readers. What happens if two sockets try to reverse the path at the same time? It's seems like an unlikely scenario at the moment but I think it would be nice if this could be fully ruled out.

Related to that, I think we discussed that before: JPAN reverses the path in-place when receiving a path. After that it is immutable. This avoids any concurrency issues and checking whether a path is already reversed or not.

tzaeschke commented on lib/data_plane/path.c:
```diff
@@ -464,3 +389,8 @@ int scion_path_deserialize(uint8_t *buf, struct scion_path_meta_hdr *hdr, struct
 	return 0;
 }
+
+void scion_path_print_metadata(struct scion_path *path)
```
`const` argument?

tzaeschke reviewed (commented):
I only spot-checked the PR. A couple of general comments:
- I just realized that default indentation is apparently 8 chars. This seems quite a lot, is this still standard today for C?
- If I remember correctly,, we said we would discuss constness later. Or did we decide to ignore it? It seems to me it may be added without too much work in same areas, for example for metadata?

koflin commented:
> * I just realized that default indentation is apparently 8 chars. This seems quite a lot, is this still standard today for C?

This is determined by the `.clang-format` config, which Marc initially added. Actually, I think we are using tabs and not spaces and according to the config the tabs should be 4 spaces wide.

> * If I remember correctly,, we said we would discuss constness later. Or did we decide to ignore it? It seems to me it may be added without too much work in same areas, for example for metadata?

I ignored it for now. But I agree that in particular places it would be quite useful to add it.

koflin commented:
@marcfrei @tzaeschke I will merge this if it looks good to you. I've created a separate issue for the path policy language #44.

marcfrei commented on lib/control_plane/path_metadata.c:
```diff
@@ -0,0 +1,720 @@
+#include "data_plane/path.h"
+#include "util/map.h"
+
+#include <assert.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
```
The block of standard headers should be included first.

marcfrei commented on lib/control_plane/policy.c:
```diff
@@ -0,0 +1,206 @@
+#include "policy.h"
+#include "common/path_collection.h"
+
+#include <stdio.h>
+#include <time.h>
```
First, include standard headers in a block, then project specific headers in a second block.

marcfrei commented on lib/control_plane/segment.c:
```diff
@@ -24,9 +24,13 @@
 #include "proto/control_plane/v1/seg.pb-c.h"
 #include "util/http2_rpc.h"
 
+#include <util/map.h>
```
Keep project specific headers in one block.

marcfrei reviewed (approved):
LGTM!

tzaeschke commented on lib/control_plane/policy.c:
```diff
@@ -0,0 +1,206 @@
+	return (hops_one > hops_two) - (hops_one < hops_two);
+}
+
+static void least_hops_filter(struct scion_path_collection *paths)
```
Apologies for this late observation:
Is it a good idea to call this a `filter`? I know JPAN does that, but I actually plan to change that.

I think there are two different operations that need to be available:
- ordering, e.g. least-hops, max bandwidth, min latency.
- filtering, e.g. minimum MTU, avoid AS by number, ....

The key difference is that the latter may remove elements from the list.

On first glance, it looks like we can implement ordering methods and filtering methods with the same API.
However, if there is a problem with how in-place operations work, then we may have to adapt the API for "ordering" vs "filtering".

So:
1) Could you implement a `min_MTU()` or a `exclude_AS()` operation just to make sure that works fine?
2) Can we rethink the naming? I think it can be confusing if we have a `min_MTU_filter` and a `highest_MTU_filter` where the first removes paths (but does no ordering) and the second orders them (but does no removals). Maybe we can use "sorter"instead of "filter" for ordering? Other suggestions?

Since this PR is already quite big we can do that in a separate PR,

koflin replied:
Yeah, I also don't like the name. I just used it to be consistent with PAN and JPAN. We could call it `filter_and_sort()` or `apply()` to be even more generic.

Both the filter and sort method on the `scion_path_collection` are in-place.

> 1. Could you implement a `min_MTU()` or a `exclude_AS()` operation just to make sure that works fine?

Sure, I'll implement a `min_MTU()`.

> 2. Can we rethink the naming? I think it can be confusing if we have a `min_MTU_filter` and a `highest_MTU_filter` where the first removes paths (but does no ordering) and the second orders them (but does no removals). Maybe we can use "sorter"instead of "filter" for ordering? Other suggestions?

This is internal naming anyways. But I can adjust it to "sort" and "filter". I wouldn't call it "sorter" because these are functions and not objects.

I actually think we should settle this in this PR, we are not in a hurry with merging this.

koflin replied:
@tzaeschke I've implemented a `scion_policy_min_mtu()` function now that returns a policy that ensures all paths have the provided minimum MTU. 

Actually, while implementing this policy generator function, I realized that we need custom contexts in policy functions too. Therefore I added a `ctx` argument to the policy function and added a `ctx` member to the `scion_policy`.

koflin replied:
This is a common pattern in C that allows a user to inject a custom context (here when creating the policy) that will be passed to the policy function when executed. It is declared `void *` because it's up to the user to decide what it should be. It is necessary for implementing **dynamic** policy functions. For instance, the policy function `filter_min_mtu()` can filter paths having a **dynamic** (meaning that it is not known at compile time) minimum MTU provided in the context.

tzaeschke replied:
Maybe it would be good to document just that? E.g. `ctx: user definable spare argument to be consumed by implementations of path filter functions` or something similar?

koflin replied:
Ah I see, I didn't document the argument for the `scion_policy_fn` type. Added it now.

tzaeschke reviewed (commented):
LGTM % One concern that could be moved to a separate PR.
